app/controllers/SocialAuthController.scala

```diff
@@ -0,0 +1,49 @@
+package controllers
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+import com.mohiva.play.silhouette.api._
+import com.mohiva.play.silhouette.api.exceptions.ProviderException
+import com.mohiva.play.silhouette.api.repositories.AuthInfoRepository
+import com.mohiva.play.silhouette.impl.providers._
+import javax.inject.Inject
+import play.api.i18n.{ I18nSupport, Messages }
+import play.api.mvc.{ AbstractController, AnyContent, ControllerComponents, Request }
+
+import models.services.UserService
+import utils.auth.DefaultEnv
+
+class SocialAuthController @Inject() (
+    components: ControllerComponents,
+    silhouette: Silhouette[DefaultEnv],
+    userService: UserService,
+    authInfoRepository: AuthInfoRepository,
+    socialProviderRegistry: SocialProviderRegistry
+  )(implicit ex: ExecutionContext)
+  extends AbstractController(components) with I18nSupport with Logger {
+
+  def authenticate(provider: String) = Action.async { implicit request: Request[AnyContent] =>
+    (socialProviderRegistry.get[SocialProvider](provider) match {
+      case Some(p: SocialProvider with CommonSocialProfileBuilder) =>
+        p.authenticate().flatMap {
+          case Left(result) => Future.successful(result)
+          case Right(authInfo) => for {
+            profile <- p.retrieveProfile(authInfo)
+            user <- userService.save(profile)
+            authInfo <- authInfoRepository.save(profile.loginInfo, authInfo)
+            authenticator <- silhouette.env.authenticatorService.create(profile.loginInfo)
+            value <- silhouette.env.authenticatorService.init(authenticator)
+            result <- silhouette.env.authenticatorService.embed(value, Redirect(routes.HomeController.google))
+          } yield {
+            silhouette.env.eventBus.publish(LoginEvent(user, request))
+            result
+          }
+        }
+      case _ => Future.failed(new ProviderException(s"Cannot authenticate with unexpected social provider $provider"))
+    }).recover {
```

Comment:
Does the above block need to be wrapped in parens to chain the `recover`? The one in `SignInController` isn't.

Reply:
tried removing, but seems like you need it. guess you can't get the result value from the block after match?